### ISSUE TITLE
Generate the icons sprite on build

### ISSUE THREAD

**Issue by nicofrand** [cozy-labs/konnectors#400]
No idea how to do it yet, just a reminder.

**Comment by ZeHiro** [comment]
have a look at Kresus sprite generation. The problem is that the lib used to generate the sprites needs compilation of C code, and can be a problem for new contributors when setting their environnement

**Comment by nicofrand** [comment]
The main issue would be to detect the icons. That would be easier to force the connector js/coffee file, the model's name and the icon file to all have the same name.

Then for each konnector we look for the icon file, read it, insert it in base64 inside the SVG, and that would be ok, no ? The question of the image's size remains but we could force it to a standard one, same for all icons ?

**Comment by nicofrand** [comment]
See an example of what I would like to implement : https://up.nicofrand.eu/test_svg/

We would need to be able to:

- know the dimensions of an image
- transform the image in base64

**Comment by ZeHiro** [comment]
@kelukelu maybe you could help us on that?

**Comment by m4dz** [comment]
Hi there!

We are thinking on how to revamp the build system and the client-side architecture. I've just made a [PR with this doc](https://github.com/cozy-labs/konnectors/pull/531/files#diff-1edf711e5468c5b76de4af01de431bad).

We'll move the build tool to Webpack, as for other Cozy's apps, and the SVG sprites is properly handled with a dedicated loader (so, it's on our roadmap :)).

**Comment by Phyks** [comment]
Why not forcing the user to put moving konnectors to subfolders (that is have a tree like "konnectors/MyKonnector/{MyKonnector.js, ...}" and expect the dev to put a "icon.svg" or "{slugified_name}.svg" with a given canvas size in this folder?

Then, writing a Node script to fetch all the available SVG files (and use a default file if none is found for a given Konnector) and concatening them all at build time sounds quite easy.

**Comment by m4dz** [comment]
I think we need to clarify for which use we need to build a sprite. @nicofrand what have you I mind originally? Is it for services logos, or for Ui icons (buttons, etc)?

**Comment by nicofrand** [comment]
For websites (materiel.net, sncf, etc.). And I had the same idea than Phyks (although we could keep the icons in a separate directory)

**Comment by ZeHiro** [comment]
well if the konnector is completely separated, the imbage should be in the konnector script/bundle.

**Comment by m4dz** [comment]
OK, I understand. My main problem with the sprite solution is that we can't be sure that all images (aka logos) will always be available in an SVG format. So what if we only have a PNG? The other way is to rely on an old type of bitmap sprite containing all logos, but it won't be a really good solution for images scaling (particularly in a responsive mode).

A third solution can be to serve each image behind a endpoint (e.g. `/konnectors/<service_name>/icon`) or directly embeded in the JSON config response. But it means a new request per icon 😕.

@nono, any thought on it?

**Comment by nicofrand** [comment]
If you build a SVG sprite with PNG icons embedded in base64 format and SVG icons as … SVG, that should be fine ?

**Comment by m4dz** [comment]
> build a SVG sprite with PNG icons embedded in base64 format and SVG icons as SVG

Fine, but ugly. But fine :)

**Comment by nicofrand** [comment]
Not that ugly…

**Comment by Phyks** [comment]
Something is quite unclear for me. Why could not we serve a "static" folder directly, rather than having to add a route for each and every icon?

**Comment by m4dz** [comment]
> Not that ugly…

Yeah, not *that* ugly, but embedding rwsterized images in SVG isn't very fair, as we expect the SVG elemens and sprites to be vectorials. But indeed, it can be a sustainable solution if icons in bitmap format are some exception.

> Why could not we serve a "static" folder directly

Here, it's more about the stack v2 as mentioned by @nono in #531: I don't know if the stack would be able to serve some arbitrary static content, nor if it's a reasonnable choice (it means that the stack whould be able to _collect_ all static items from various submodules to serves them in a unified way in a static dir. My opinion is that using a dedicated endpoint to serve those knd of assets is a more maintainable way to do it, but I could be wrong.

**Comment by Phyks** [comment]
> Here, it's more about the stack v2 as mentioned by @nono in #531: I don't know if the stack would be able to serve some arbitrary static content, nor if it's a reasonnable choice (it means that the stack whould be able to collect all static items from various submodules to serves them in a unified way in a static dir. My opinion is that using a dedicated endpoint to serve those knd of assets is a more maintainable way to do it, but I could be wrong.

Sorry, I missed it in #531 I think. What I am used to when writing such routers is to basically have a "/static" directory somewhere, containing all the static files to be served and serve this whole folder as an endpoint.

Typically, we could have `/static/konnectors/KONNECTOR_NAME.{svg,png}` files and just serve `/static/<:path>` endpoint? (What I have in mind is somewhat similar to [what is done with Bottle](http://bottlepy.org/docs/dev/tutorial.html#routing-static-files) although we are not under the exact same pattern.

**Comment by m4dz** [comment]
Yep, I have something similar in mind with Django and its `collectstatic` command :)

I let @nono reply to it, he probably has some good solution about this one, maybe using the File service API as described in https://github.com/cozy/cozy-stack/blob/master/docs/architecture.md

**Comment by Phyks** [comment]
Ok :) To sum up, options for now seems to be:

* Build a sprite at build time as should be done right now and serve this unique sprite.
* Use a `collectstatic`-like mechanism and serve all static files in a given folder.
* Put base64 encoded icons in the config JSON.

**Comment by ZeHiro** [comment]
As the konnectors and apps will be in different repo, serving one icon per konnector could be interesting (no sprite build to be done in any repo). Or have a simple way to provide a new sprite

**Comment by nono** [comment]
> I don't know if the stack would be able to serve some arbitrary static content

The stack will be able to do that.

**Comment by Phyks** [comment]
After discussing with @m4dz on IRC, it seems passing icons in JSON config file directly could be a sensible option. Indeed, images are quite small, and most of them will be SVG files, so there should not be much overhead in passing them as data URI in the JSON config file. Moreover, it should be nicely gzipped. It would also limit the number of requests to be made as icons will always be fetched.

The only issue I could think of about this method is that it prevents the client from caching the icons.

**Comment by m4dz** [comment]
@Phyks 😚 👌
